pkg/cmd/roachtest/prometheus.go

```diff
@@ -0,0 +1,166 @@
+// Copyright 2021 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/option"
+	"gopkg.in/yaml.v2"
+)
+
+// prometheusScrapeNode are nodes to scrape from.
+type prometheusScrapeNode struct {
+	node option.NodeListOption
+	port int
+}
+
+// prometheusConfig is a monitor that watches over the running of prometheus.
+type prometheusConfig struct {
+	prometheusNode option.NodeListOption
+	jobName        string
+	scrapeNodes    []prometheusScrapeNode
+}
+
+// run installs and runs prometheus.
+// Prometheus will run in the background and log an error if it fails, as opposed
+// to failing the test.
+func (pm *prometheusConfig) run(ctx context.Context, t *test, c Cluster) {
+	cfg, err := makePrometheusConfig(
+		ctx,
+		c,
+		pm.jobName,
+		pm.scrapeNodes,
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if err := repeatRunE(
+		ctx,
+		t,
+		c,
+		pm.prometheusNode,
+		"install prometheus",
+		fmt.Sprintf(
+			`rm -rf /tmp/prometheus && mkdir /tmp/prometheus && cd /tmp/prometheus &&
+			wget -O prometheus.tar.gz https://github.com/prometheus/prometheus/releases/download/v2.27.1/prometheus-2.27.1.linux-amd64.tar.gz &&
+			tar xvf prometheus.tar.gz --strip-components=1 &&
+			echo "%s" > prometheus.yml
+		`,
+			cfg,
+		),
+	); err != nil {
+		t.Fatal(err)
+	}
+
+	// Start prometheus in an async thread.
+	go func() {
+		if err := c.RunE(
+			ctx,
+			pm.prometheusNode,
+			`cd /tmp/prometheus &&
+			./prometheus --config.file=prometheus.yml --storage.tsdb.path=data/ --web.enable-admin-api`,
```

Comment:
🤔 We've recently converted our cockroach startup to use systemd-run. I wonder if it would be worth it to do that here as well.

Reply:
Yes, really good idea. We can `PutString` the entry point script (if we even need one) to keep this file clean.
Ultimately, we want everything to be launched on these clusters through systemd. While that's a pipedream as of yet (due to the large numbers of one-off commands) it is where we should be heading.

Reply:
The existing systemd code is here: https://github.com/cockroachdb/cockroach/blob/98a75bb3b9ad10c27336fbe76c92489b4cdbf6f2/pkg/cmd/roachprod/install/cockroach.go#L378-L458

The self-installing re-entrant bash script pattern is certainly not the part to copy here. Just upload a script (if needed), install the job, start the job.